File: SimpleWhpDemo/cga.c
#include "cga.h"
#ifndef _WIN32
#include <sys/time.h>
#endif

UCHAR CgaMode = 0;
UCHAR AttrCga = 0;
UCHAR CgaStatus = 0;
ULONGLONG CgaLastToggleMs = 0;
UCHAR CrtcCgaIndex = 0;
UCHAR CrtcCgaRegs[32] = {0};

void CgaOut(USHORT port, UCHAR val)
{
    switch (port)
    {
    case 0x3D4:
    case 0x3D6:
        CrtcCgaIndex = val & 0x1F;
        break;
    case 0x3D5:
    case 0x3D7:
        CrtcCgaRegs[CrtcCgaIndex] = val;
        break;
    case 0x3D8:
        CgaMode = val;
        break;
    case 0x3D9:
        AttrCga = val;
        break;
    default:
        break;
    }
}

static ULONGLONG cga_now_ms(void)
{
#ifdef _WIN32
    return GetTickCount64();
#else
    struct timeval tv;
    gettimeofday(&tv, NULL);
    return (ULONGLONG)tv.tv_sec * 1000ULL + tv.tv_usec / 1000ULL;
#endif
}

void CgaInit(void)
{
    CgaMode = AttrCga = 0;
    CgaStatus = 0;
    CgaLastToggleMs = cga_now_ms();
}

UCHAR CgaIn(USHORT port)
{
    if (port == 0x3DA)
    {
        ULONGLONG now = cga_now_ms();
        /* Simulate display enable toggling at ~60 Hz */
        if (now - CgaLastToggleMs >= 16)
        {
            CgaStatus ^= 0x01;
            CgaLastToggleMs += 16;
        }
        /* Keep vertical retrace bit simple: toggle every 1 ms */
        if (((now / 1) & 1) != 0)
            CgaStatus |= 0x08;
        else
            CgaStatus &= ~0x08;

        return CgaStatus;
    }
    else if (port == 0x3D4)
    {
        return CrtcCgaIndex;
    }
    else if (port == 0x3D5)
    {
        return CrtcCgaRegs[CrtcCgaIndex];
    }
    return 0xFF;
}

File: src/cga.rs
pub static mut MODE: u8 = 0;
pub static mut COLOR: u8 = 0;
pub static mut CRTC_INDEX: u8 = 0;
pub static mut CRTC_REGS: [u8; 32] = [0; 32];
pub static mut STATUS: u8 = 0;
pub static mut LAST_TOGGLE: Option<std::time::Instant> = None;
const TOGGLE_PERIOD: std::time::Duration = std::time::Duration::from_millis(16);

pub fn cga_out(port: u16, val: u8) {
    unsafe {
        match port {
            0x3D4 | 0x3D6 => CRTC_INDEX = val & 0x1F,
            0x3D5 | 0x3D7 => CRTC_REGS[CRTC_INDEX as usize] = val,
            0x3D8 => MODE = val,
            0x3D9 => COLOR = val,
            _ => {}
        }
    }
}

pub fn cga_in(port: u16) -> u8 {
    unsafe {
        match port {
            0x3D4 => CRTC_INDEX,
            0x3D5 => CRTC_REGS[CRTC_INDEX as usize],
            0x3DA => {
                let now = std::time::Instant::now();
                if let Some(last) = LAST_TOGGLE {
                    if now.duration_since(last) >= TOGGLE_PERIOD {
                        STATUS ^= 0x08;
                        LAST_TOGGLE = Some(now);
                    }
                } else {
                    LAST_TOGGLE = Some(now);
                }
                STATUS
            }
            _ => 0xFF,
        }
    }
}
